File: compare_org_local.js
const fs = require('fs');
const path = require('path');

function getLocalMetadata(forceAppDir) {
    const localMeta = {};
    const folderToType = {
        'classes': 'ApexClass',
        'pages': 'ApexPage',
        'components': 'ApexComponent',
        'triggers': 'ApexTrigger',
        'lwc': 'LightningComponentBundle',
        'aura': 'AuraDefinitionBundle',
        'objects': 'CustomObject',
        'layouts': 'Layout',
        'permissionsets': 'PermissionSet',
        'profiles': 'Profile',
        'tabs': 'CustomTab',
        'applications': 'CustomApplication',
        'flexipages': 'FlexiPage',
        'flow': 'Flow',
        'flows': 'Flow',
        'globalValueSets': 'GlobalValueSet',
        'standardValueSets': 'StandardValueSet',
        'labels': 'CustomLabels',
        'staticresources': 'StaticResource'
    };

    function traverse(dir) {
        if (!fs.existsSync(dir)) return;
        const files = fs.readdirSync(dir);
        for (const file of files) {
            const fullPath = path.join(dir, file);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory()) {
                const parts = path.relative(forceAppDir, fullPath).split(path.sep);
                const baseFolder = parts[0];
                const metaType = folderToType[baseFolder];
                
                if (metaType === 'CustomObject') {
                    if (parts.length === 2) {
                        localMeta['CustomObject'] = localMeta['CustomObject'] || new Set();
                        localMeta['CustomObject'].add(parts[1]);
                    } else if (parts.length === 3 && parts[2] === 'fields') {
                        traverse(fullPath);
                    }
                } else if (metaType === 'LightningComponentBundle' || metaType === 'AuraDefinitionBundle') {
                    if (parts.length === 2) {
                        localMeta[metaType] = localMeta[metaType] || new Set();
                        localMeta[metaType].add(parts[1]);
                    }
                } else {
                    traverse(fullPath);
                }
            } else {
                const parts = path.relative(forceAppDir, fullPath).split(path.sep);
                const baseFolder = parts[0];
                const metaType = folderToType[baseFolder];
                
                if (metaType === 'CustomObject' && parts.length === 4 && parts[2] === 'fields') {
                    if (file.endsWith('.field-meta.xml')) {
                        const objName = parts[1];
                        const fieldName = file.replace('.field-meta.xml', '');
                        localMeta['CustomField'] = localMeta['CustomField'] || new Set();
                        localMeta['CustomField'].add(`${objName}.${fieldName}`);
                    }
                } else if (metaType && metaType !== 'LightningComponentBundle' && metaType !== 'AuraDefinitionBundle' && metaType !== 'CustomObject') {
                    if (file.endsWith('-meta.xml') || file.endsWith('.xml')) {
                        let name = file.split('.')[0];
                        if (name) {
                            localMeta[metaType] = localMeta[metaType] || new Set();
                            localMeta[metaType].add(name);
                        }
                    } else if (file.endsWith('.cls') && metaType === 'ApexClass') {
                        localMeta[metaType] = localMeta[metaType] || new Set();
                        localMeta[metaType].add(file.replace('.cls', ''));
                    }
                }
            }
        }
    }
    traverse(forceAppDir);
    return localMeta;
}

function getOrgMetadata(manifestPath) {
    const orgMeta = {};
    if (!fs.existsSync(manifestPath)) return orgMeta;
    
    const content = fs.readFileSync(manifestPath, 'utf8');
    const typesRegex = /<types>([\s\S]*?)<\/types>/g;
    let match;
    
    while ((match = typesRegex.exec(content)) !== null) {
        const typeBlock = match[1];
        const nameMatch = /<name>(.*?)<\/name>/.exec(typeBlock);
        if (nameMatch) {
            const metaType = nameMatch[1];
            orgMeta[metaType] = new Set();
            
            const memberRegex = /<members>(.*?)<\/members>/g;
            let memberMatch;
            while ((memberMatch = memberRegex.exec(typeBlock)) !== null) {
                orgMeta[metaType].add(memberMatch[1]);
            }
        }
    }
    return orgMeta;
}

function main() {
    const localDir = 'C:/Users/crist/OneDrive/Documentos/Proyecto pyatz/Pyatz/force-app/main/default';
    const manifestPath = 'C:/Users/crist/OneDrive/Documentos/Proyecto pyatz/Pyatz/manifest/org_manifest.xml';
    
    const local = getLocalMetadata(localDir);
    const org = getOrgMetadata(manifestPath);
    
    console.log("=== DISCREPANCY ANALYSIS ===");
    
    const typesToCheck = ['ApexClass', 'LightningComponentBundle', 'CustomObject', 'CustomField', 'ApexTrigger', 'FlexiPage', 'Flow'];
    
    for (const t of typesToCheck) {
        console.log(`\n--- ${t} ---`);
        const orgSet = org[t] || new Set();
        const localSet = local[t] || new Set();
        
        const inOrgNotLocal = new Set([...orgSet].filter(x => !localSet.has(x)));
        // Ignore managed packages with namespaces (usually containing __ in the prefix, but not __c)
        const inOrgNotLocalFiltered = [...inOrgNotLocal].filter(x => !(x.includes('__') && !x.endsWith('__c') && !x.endsWith('__mdt')));
        
        if (inOrgNotLocalFiltered.length > 0) {
            console.log(`[${inOrgNotLocalFiltered.length}] In Org ONLY (missing locally):`);
            inOrgNotLocalFiltered.sort().slice(0, 30).forEach(item => console.log(`  + ${item}`));
            if (inOrgNotLocalFiltered.length > 30) {
                console.log(`  ... and ${inOrgNotLocalFiltered.length - 30} more`);
            }
        }
        
        const inLocalNotOrg = [...localSet].filter(x => !orgSet.has(x));
        if (inLocalNotOrg.length > 0) {
            console.log(`[${inLocalNotOrg.length}] In Local ONLY (not deployed to org):`);
            inLocalNotOrg.sort().slice(0, 30).forEach(item => console.log(`  - ${item}`));
            if (inLocalNotOrg.length > 30) {
                console.log(`  ... and ${inLocalNotOrg.length - 30} more`);
            }
        }
        
        if (inOrgNotLocalFiltered.length === 0 && inLocalNotOrg.length === 0) {
            console.log("  ✓ Perfectly synced");
        }
    }
}

main();

File: compare_org_local.py
import os
import xml.etree.ElementTree as ET
from collections import defaultdict
import glob

def get_local_metadata(force_app_dir):
    local_meta = defaultdict(set)
    
    # Simple mapping based on folder names
    # Mapping folder names to XML names (approximate for most common types)
    folder_to_type = {
        'classes': 'ApexClass',
        'pages': 'ApexPage',
        'components': 'ApexComponent',
        'triggers': 'ApexTrigger',
        'lwc': 'LightningComponentBundle',
        'aura': 'AuraDefinitionBundle',
        'objects': 'CustomObject',
        'layouts': 'Layout',
        'permissionsets': 'PermissionSet',
        'profiles': 'Profile',
        'tabs': 'CustomTab',
        'applications': 'CustomApplication',
        'flexipages': 'FlexiPage',
        'flow': 'Flow', # sometimes flows
        'flows': 'Flow',
        'globalValueSets': 'GlobalValueSet',
        'standardValueSets': 'StandardValueSet',
        'labels': 'CustomLabels', # special case
        'staticresources': 'StaticResource'
    }

    # Iterate through folders
    for root, dirs, files in os.walk(force_app_dir):
        # We only care about main folders under main/default
        rel_path = os.path.relpath(root, force_app_dir)
        parts = rel_path.split(os.sep)
        
        if len(parts) >= 1 and parts[0] in folder_to_type:
            meta_type = folder_to_type[parts[0]]
            
            # For objects, we also need to get fields
            if meta_type == 'CustomObject':
                if len(parts) == 1:
                    for d in dirs:
                        local_meta['CustomObject'].add(d)
                elif len(parts) == 3 and parts[2] == 'fields':
                    obj_name = parts[1]
                    for f in files:
                        if f.endswith('.field-meta.xml'):
                            local_meta['CustomField'].add(f"{obj_name}.{f.replace('.field-meta.xml', '')}")
            elif meta_type == 'LightningComponentBundle' or meta_type == 'AuraDefinitionBundle':
                if len(parts) == 1:
                    for d in dirs:
                        local_meta[meta_type].add(d)
            else:
                for f in files:
                    if f.endswith('-meta.xml') or f.endswith('.xml'):
                        # Exclude the -meta.xml part
                        name = f.split('.')[0]
                        if name:
                            local_meta[meta_type].add(name)
                            
    return local_meta

def get_org_metadata(manifest_path):
    org_meta = defaultdict(set)
    if not os.path.exists(manifest_path):
        print(f"Error: {manifest_path} not found")
        return org_meta
        
    tree = ET.parse(manifest_path)
    root = tree.getroot()
    # Namespace is usually xmlns="http://soap.sforce.com/2006/04/metadata"
    ns = {'ns': 'http://soap.sforce.com/2006/04/metadata'}
    
    for types in root.findall('ns:types', ns):
        name_node = types.find('ns:name', ns)
        if name_node is not None:
            meta_type = name_node.text
            for member in types.findall('ns:members', ns):
                org_meta[meta_type].add(member.text)
                
    return org_meta

def main():
    local_dir = 'C:/Users/crist/OneDrive/Documentos/Proyecto pyatz/Pyatz/force-app/main/default'
    manifest_path = 'C:/Users/crist/OneDrive/Documentos/Proyecto pyatz/Pyatz/manifest/org_manifest.xml'
    
    local = get_local_metadata(local_dir)
    org = get_org_metadata(manifest_path)
    
    print("=== DISCREPANCY ANALYSIS ===")
    
    types_to_check = ['ApexClass', 'LightningComponentBundle', 'CustomObject', 'CustomField', 'ApexTrigger', 'FlexiPage', 'Flow']
    
    for t in types_to_check:
        print(f"\\n--- {t} ---")
        org_set = org.get(t, set())
        local_set = local.get(t, set())
        
        # In Org but NOT Local
        in_org_not_local = org_set - local_set
        # Ignore managed packages with namespaces (usually containing __ in the prefix, but not __c)
        in_org_not_local = {x for x in in_org_not_local if not ('__' in x and not x.endswith('__c'))}
        
        if in_org_not_local:
            print(f"[{len(in_org_not_local)}] In Org ONLY (missing locally):")
            for item in sorted(list(in_org_not_local))[:20]:
                print(f"  + {item}")
            if len(in_org_not_local) > 20:
                print(f"  ... and {len(in_org_not_local) - 20} more")
                
        in_local_not_org = local_set - org_set
        if in_local_not_org:
            print(f"[{len(in_local_not_org)}] In Local ONLY (not deployed to org):")
            for item in sorted(list(in_local_not_org))[:20]:
                print(f"  - {item}")
            if len(in_local_not_org) > 20:
                print(f"  ... and {len(in_local_not_org) - 20} more")
                
        if not in_org_not_local and not in_local_not_org:
            print("  ✓ Perfectly synced")

if __name__ == '__main__':
    main()
